doc/configuration.md
# Configuration

There are several entry points available for administrator to manage the configuration of Universal Database. This section aims to outline those configurations and explain each option available and what it does.

You may configure every option using the configuration file, command line argument or environment variable.

Take note that configuration options are distinct from the runtime setting, available from the web interface. The configuration options here usually meant to be static and set before starting the server. You may get the list of configuration options by calling `udb --help`.

Note: If an option is specified in more than one place, the command line arguments override the environment variable, environment variables override config files, and config files override default value.

## Configuration file

To use configuration files, you may call `udb` with `-f` or `--config` to define the configuration file location. When not defined, Universal Database loads all configuration files from these locations by default:

* /etc/udb/udb.conf
* /etc/udb/udb.conf.d/*.conf

Configuration file syntax must define a key and a value. The key is case-sensitive, and you may use underscore (_) or dash (-) seemlessly. All lines beginning with '#' are comments and are intended for you to read. All other lines are configuration for udb.

E.g.:

    # This is a comment
    server_port=8081
    log_level=DEBUG

## Environment variables

In addition to configuration files, you may pass environment variables. The options name must be uppercase and prefixed with `UDB_`. As an example, if you want to change the port used to listen for HTTP request for 8081, you must define `server-port` option as follow.

    UDB_SERVER_PORT=8081

## Command line arguments

When launching `udb` executable, you may pass as many arguments as you want on the command line. The options must be prefixed with double dash (`--`) and you must single dash (-) to separate words.

E.g. `--server-port 8081` or `--server-port=8081` are valid

## Configure listening port and interface

For security reasons, Universal Database listen on port `8080` for HTTP request on loopback interface (127.0.0.1) by default. Consider configuring a reverse proxy like Nginx or Apache2 if you want to make Universal Database listen on port 80 for HTTP and port 443 for HTTPS request.

| Option | Description | Example |
| --- | --- | --- |
| server-host | Define the IP address to listen to. Use `0.0.0.0` to listen on all interfaces. Use `127.0.0.1` to listen on loopback interface. | 0.0.0.0 |
| server-port | Define the port to listen for HTTP request. Default to `8080` | 9090 |

## Configure External URL

To display the correct URL when sending email notification to UDB users,
you must provide it with the URL your users use to reach the web application.
You can use the IP of your server, but a Fully Qualified Domain Name (FQDN) is preferred.

| Option | Description | Example |
| --- | --- | --- |
| external-url | Define the base URL used to reach your Universal Database application | `https://udb.mycompagny.com` |

## Configure administrator username and password

Using configuration file, you may setup a special administrator which cannot be
deleted or renamed from the web interface. You may also configure a specific
password for this user that cannot be updated from the web interface either.

In addition, you may also create other administrator users to manage Universal Database.

| Parameter | Description | Example |
| --- | --- | --- |
| admin-user | Define the name of the default admin user to be created | admin |
| admin-password | administrator encrypted password as SSHA. Read online documentation to know more about how to encrypt your password into SSHA or use <http://projects.marsching.org/weave4j/util/genpassword.php> When defined, administrator password cannot be updated using the web interface. When undefined, default administrator password is `admin123` and it can be updated using the web interface. | modification |

## Configure logging

Universal Database can be configured to send logs to specific location. By default, logs are sent to the console (stdout or stderr). If you have installed Universal Database on a server, you should consider enabling the logging to help you keep track of the activities or to help you debug problem.

| Option | Description | Example |
| --- | --- | --- |
| log-level | Define the log level. ERROR, WARN, INFO, DEBUG | DEBUG |
| log-file | Define the location of the log file. | /var/log/udb/server.log |
| log-access-file | Define the location of the access log file. | /var/log/udb/access.log |

### Enable Debugging

A specific option is also available if you want to enable the debugging log. We do not recommend to enable this option in production as it may leak information to the user whenever an exception is raised.

| Option | Description |
| --- | --- |
| debug | enable UDB debug mode - change the log level to DEBUG, print exception stack trace to the web interface and show SQL query in logs. |

## Configure database

Universal Database use SQL database to store user preferences. The embedded SQLite database is well suited for small deployment (1-100 users). If you intended to have a large deployment, you must consider using a PostgreSQL database instead.

| Option | Description | Example |
| --- | --- | --- |
| database-uri | Location of the database used for persistence. SQLite and PostgreSQL database are supported officially. To use a SQLite database, you may define the location using a file path or a URI. e.g.: `/srv/udb/file.db` or `sqlite:///srv/udb/file.db`. To use PostgreSQL server, you must provide a URI similar to `postgresql://user:pass@10.255.1.34/dbname` and you must install required dependencies. By default, Universal Database uses a SQLite embedded database located at `/etc/udb/rdw.db`. | postgresql://user:pass@10.255.1.34/dbname |

### SQLite

To use embedded SQLite database, pass the option `database-uri` with a URI similar to `sqlite:///etc/udb/data.db` or `/etc/udb/data.db`.

### PostgreSQL

To use an external PostgreSQL database, pass the option `database-uri` with a URI similar to `postgresql://user:pass@10.255.1.34/dbname`.

You may need to install additional dependencies to connect to PostgreSQL. Step to install dependencies might differ according to the way you installed Universal Database.

**Using Debian repository:**

    apt install python3-psycopg2

**Using Pypi repository:**

    pip install psycopg2-binary

## Configure LDAP Authentication

Universal Database may integrates with LDAP server to support user authentication.

```{toctree}
---
titlesonly: true
---
configuration-ldap
```

## Configure User Session

A user session is a sequence of request and response transactions associated with a single user. The user session is the means to track each authenticated user.

| Option | Description | Example |
| --- | --- | --- |
| session-idle-timeout | This timeout defines the amount of time a session will remain active in case there is no activity in the session. User Session will be revoke after this period of inactivity, unless the user selected "remember me". Default 10 minutes. | 5 |
| session-absolute-timeout | This timeout defines the maximum amount of time a session can be active. After this period, user is forced to (re)authenticate, unless the user selected "remember me". Default 20 minutes. | 30 |
| session-persistent-timeout | This timeout defines the maximum amount of time to remember and trust a user device. This timeout is used when user select "remember me". Default 30 days | 43200 |

## Configure email notifications

You may configure Universal Database to send an email notification to the users for record they follow.

| Option | Description | Example |
| --- | --- | --- |
| smtp-encryption | Type of encryption to be used when establishing communication with SMTP server. Available values: `none`, `ssl` and `starttls` | starttls |
| smtp-server | SMTP server used to send email in the form `host`:`port`. If the port is not provided, default to standard port 25 or 465 is used. | smtp.gmail.com:587 |
| smtp-from | email addres used for the `From:` field when sending email. | Universal Database <example@gmail.com> |
| smtp-username | username used for authentication with the SMTP server. | example@gmail.com |
| smtp-password | password used for authentication with the SMTP server. | CHANGEME |
| notification-catch-all-email | When defined, all notification email will be sent to this email address using Blind carbon copy (Bcc) |

To configure the notification, you need a valid SMTP server. In this example, you are making use of a Gmail account to send emails.

    smtp-server=smtp.gmail.com:587
    smtp-encryption=starttls
    smtp-from=example@gmail.com
    smtp-username=example@gmail.com
    smtp-password=CHANGEME

Note: notifications are not sent if the user doesn't have an email configured in his profile.

## Configure Rate-Limit

Universal Database could be configured to rate-limit access to anonymous to avoid bruteforce
attacks and authenticated users to avoid Denial Of Service attack.

| Option | Description | Example |
| --- | --- | --- |
| rate-limit | maximum number of requests per hour that can be made on sensitive endpoints. When this limit is reached, an HTTP 429 message is returned to the user or the user is logged out. This security measure is used to limit brute force attacks on the login page and the RESTful API. | 20 |
| rate-limit-dir | location where to store rate-limit information. When undefined, data is kept in memory. | /var/lib/udb/session |

## Custom user's password length limits

By default, Universal Database encofrce minimal password score of 2.

Changing the minimum score does not affect existing users' passwords. Existing users are not prompted to reset their passwords to meet the new limits. The new limit only applies when an existing user changes their password.

| Option | Description | Example |
| --- | --- | --- |
| password-score      | Minimum zxcvbn's score for password. Value from 0 to 4. Default value 2. | 4 |

You may want to read more about [zxcvbn](https://github.com/dropbox/zxcvbn) score value.

## Configure default language

By default, the web application uses the HTTP Accept-Language headers to determine the best language to use for display. Users can also manually select a preferred language to use for all communication. The `default-lang` setting is used when the user has not selected a preferred language and none of the Accept-Language headers match a translation.

| Parameter | Description | Example |
| --- | --- | --- |
| --default-lang | default application locale. e.g.: `fr` | es |

## Configure default timezone

By default, the web application uses the browse timezone to display date and time. But when sending email, the server timezone is used by default. It's possible for the administrator to change the default timezone to be used.

It's also possible for users to changed their prefered timezone from the web interface.

| Parameter | Description | Example |
| --- | --- | --- |
| --default-timezone | default application timezone. uses server timezone by default. | America/toronto |


CHANGELOG.md
# Change Log

## Next Release

* Enforce 'Origin' validation to counter CSRF and XSS
* Add ratelimit protection on sensitive endpoints
* Enforce strong password with zxcvbn
* Enforce field limit in wtform validation
* Implement POST, REDIRECT, GET in user profile
* Update Content-Security-Policy
* Update auth_form
* Remove deprecated auth_basic module
* Define proper proxy settings for best security #67
* Avoid leaking stacktrace when running in production mode #67
* Define default security headers #67
* Redirect user to same page after edit
* Complete datatable changes
* Support VRF Records
* Enabled `cherrypy.tools.proxy` to support reverse proxy #65
* Use datatables library to display tables #48 #66
* Provide default log configuration for debian package
* Change package name
* Update db module plugins
* Cosmetic changes
* Allow Administrators to define or remove user's password #51
* Provide a default page for /api/
* Provide a default configfile #52
* Update db Base class
* Make IP View a custom template
* Ad scalar_subquery to remove sqlalchemy warning
* Redirect user to previous page after editing #22
* Define inherit_cache for cidr
* Fix URL in notifications view
* Define default image to upload deb
* Define a timestamp decorator to handle timezone
* Add a dashboard page #26
* Add related_subnets to IP View #39
* Add logo to login page
* Add `cc` to smtp module
* Avoid editing status using HTTP GET method
* Avoid editing followers using HTTP GET
* Fix open redirect in login page #45
* Add .vscode to gitignore
* Validate PTR Record using DNS Zone #37
* Reffactoring reverse_ip as hybrid_property
* Exclude deleted records from relation to IP #42
* Exclude generated templates from sonar
* Fix to support email-validator v1.2.0
* Move display_name to jinja2 templates
* Add a view to list all subscriptions #40
* Add catch-all email to configuration #44
* Add alt to image logo
* Make email template compliant
* Add aria-describedby to table
* Add aria-hidden="true" to icons
* Fix notification
* Add catch-all email to configuration #44
* Fetch email and fullname attributes from LDAP #43
* Send notification to followers #15
* Update Copyright year
* Replace font awsome by bootstrap-icons #24 #34
* Fix sonar coverage
* Add full text search #28
* Fix test for postgresql database
* Make tables sortable #30
* Run test during Debian Packaging using pytest
* Add DNS Zone / Subnet / Dns Record validation #21
* Validate if subnet creation conflict with other existing subnet #9
* Add isort and black to project
* Configuring setup() using setup.cfg files
* Update sidebar
* Complete user managements
* Implement LDAP authentication #4
* Add new plugins to the project
* Add user managements #27
* Include PTR record ipv4 and ipv6 in IP View
* Make database-uri configurable #25 #33
* Add specific PTR record validation #31
* Add audit log on record creation #29
* Complete IP View changes #7
* Provide a Debian package #14
* Add IP View with related records #7
* Add all the required DNS record type #19
* Add Many-to-Many dnszone and subnet #18
* Add basic favicon #11
* Fix history when owner is updated #13
* Use a single table for messages
* Use a single table for followers
* Adjust CSS Layout
* Fix "Not Assigned" owner #20
* Fix enabled state #17
* Rename application from CMDB to UDB
* Creatre RESTful API for CRUD operation #8
* Working version
* Re-implement navigation using sidebar
* Adding more unit test
* Add more record type with test and validation
* Implement basic view to add,edit record
* Update SQLAlchemy integration
* Complete project skeleton & creating login page #3
* Fix test error and flake8


src/udb/config.py
# -*- coding: utf-8 -*-
# udb, A web interface to manage IT network
# Copyright (C) 2022 IKUS Software inc.
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse
import grp
import pwd
import sys

import configargparse

from udb.tools.i18n import gettext as _

# Get package version
try:
    import pkg_resources

    __version__ = pkg_resources.get_distribution("udb").version
except Exception:
    __version__ = 'DEV'


def _userid(value):
    """
    Parse the user value attribute that could be either a username or a userid.
    """
    try:
        return int(value)
    except ValueError:
        pass
    try:
        return pwd.getpwnam(value).pw_uid
    except KeyError:
        raise argparse.ArgumentError('user %s not found' % value)


def _groupid(value):
    """
    Parse the group value that could be either a group name or groupid.
    """
    try:
        return int(value)
    except ValueError:
        pass
    try:
        return grp.getgrnam(value).gr_gid
    except KeyError:
        raise argparse.ArgumentError('group %s not found' % value)


def _umask(value):
    """
    Validate the umask value. Read it as octal and return integer.
    """
    try:
        return int(value, base=8)
    except ValueError:
        raise argparse.ArgumentError('invalid umask value %s' % value)


def parse_args(args=None, config_file_contents=None):
    """
    Load application configuration using program's arguments or environment variables.
    """
    args = sys.argv[1:] if args is None else args

    parser = configargparse.ArgumentParser(
        description='A web interface to manage IT network',
        default_config_files=['/etc/udb/udb.conf', '/etc/udb/udb.conf.d/*.conf'],
        add_env_var_help=True,
        auto_env_var_prefix='UDB_',
    )
    parser.add_argument(
        '-d',
        '--debug',
        action='store_true',
        help=_('enable debug mode - change the log level to DEBUG, print exception stack trace to the web interface'),
    )
    parser.add_argument('-f', '--config', is_config_file=True, help=_('configuration file path'))
    parser.add_argument('-v', '--version', action='version', version='udb ' + __version__)
    parser.add_argument(
        '--server-host',
        metavar='IP',
        help=_('Define the IP address to listen to. Default: 127.0.0.1'),
        default='127.0.0.1',
    )
    parser.add_argument(
        '--server-port', metavar='PORT', help=_('Define the port to listen to. Default: 8080'), default='8080', type=int
    )
    parser.add(
        '--external-url',
        metavar='URL',
        help='URL that should be used to reach this service. You can use the IP of your server, but a Fully Qualified Domain Name (FQDN) is preferred.',
    )
    parser.add_argument('--log-file', metavar='FILE', help=_('Define the location of the log file.'), default='')
    parser.add_argument(
        '--log-access-file', metavar='FILE', help=_('Define the location of the access log file.'), default=''
    )
    parser.add_argument(
        '--log-level',
        '--loglevel',
        help=_('Define the log level.'),
        choices=['ERROR', 'WARN', 'INFO', 'DEBUG'],
        default='INFO',
    )

    parser.add_argument(
        '--default-lang',
        help='default application locale. e.g.: `fr`',
        default='en_US',
    )

    parser.add_argument(
        '--default-timezone',
        help='default application timezone. e.g.: `America/toronto` uses server timezone by default.',
    )

    parser.add(
        '--session-dir',
        '--sessiondir',
        '--rate-limit-dir',
        metavar='FOLDER',
        help=_(
            'location where to store user session information and rate-limit information. When undefined, the data are kept in memory.'
        ),
    )

    parser.add(
        '--rate-limit',
        metavar='LIMIT',
        type=int,
        default=20,
        help=_(
            'maximum number of requests per hours that can be made on sensitive endpoint. When this limit is reached, an HTTP 429 message is returned to the user or user get logged out. This security measure is used to limit brute force attacks on the login page and the RESTful API. Default: 20'
        ),
    )

    parser.add(
        '--session-idle-timeout',
        metavar='MINUTES',
        help='This timeout defines the amount of time a session will remain active in case there is no activity in the session. User Session will be revoke after this period of inactivity, unless the user selected "remember me". Default 5 minutes.',
        default=10,
    )

    parser.add(
        '--session-absolute-timeout',
        metavar='MINUTES',
        help='This timeout defines the maximum amount of time a session can be active. After this period, user is forced to (re)authenticate, unless the user selected "remember me". Default 20 minutes.',
        default=30,
    )

    parser.add(
        '--session-persistent-timeout',
        metavar='MINUTES',
        help='This timeout defines the maximum amount of time to remember and trust a user device. This timeout is used when user select "remember me". Default 30 days.',
        default=43200,
    )

    # Database
    parser.add(
        '--database-uri',
        metavar='URI',
        help=_(
            """Location of the database used for persistence. SQLite and PostgreSQL
            database are supported officially. To use a SQLite database you may
            define the location using a file path or a URI.
            e.g.: /srv/udb/file.db or sqlite:///srv/udb/file.db`.
            To use PostgreSQL server you must provide
            a URI similar to postgresql://user:pass@10.255.1.34/dbname and you
            must install required dependencies.
            By default, UDB uses a SQLite embedded database located at ./data.db"""
        ),
        default='sqlite:///data.db',
    )

    # Admin user
    parser.add_argument(
        '--admin-user',
        '--adminuser',
        metavar='USERNAME',
        help=_('administrator username. The administrator user get created on startup if the database is empty.'),
        default='admin',
    )

    parser.add_argument(
        '--admin-password',
        metavar='USERNAME',
        help=_(
            """administrator encrypted password as SSHA. Read online
            documentation to know more about how to encrypt your password
            into SSHA or use http://projects.marsching.org/weave4j/util/genpassword.php
            When defined, administrator password cannot be updated using the web interface.
            When undefined, default administrator password is `admin123` and
            it can be updated using the web interface."""
        ),
    )

    parser.add_argument(
        '--add-missing-user',
        '--addmissinguser',
        action='store_true',
        help=_('enable creation of users from LDAP when the credential are valid.'),
        default=False,
    )

    parser.add_argument(
        '--add-user-default-role',
        help=_(
            'default role used when creating users from LDAP. This parameter is only useful when `--add-missing-user` is enabled.'
        ),
        default='guest',
        choices=['admin', 'user', 'guest'],
    )

    # LDAP
    parser.add_argument(
        '--ldap-uri',
        help=_('URL to the LDAP server used to validate user credentials. e.g.: ldap://localhost:389'),
    )

    parser.add_argument(
        '--ldap-base-dn',
        '--ldapbasedn',
        metavar='DN',
        help=_('DN of the branch of the directory where all searches should start from. e.g.: dc=my,dc=domain'),
        default="",
    )

    parser.add_argument(
        '--ldap-scope',
        help=_('scope of the search. Can be either base, onelevel or subtree'),
        choices=['base', 'onelevel', 'subtree'],
        default="subtree",
    )

    parser.add_argument('--ldap-tls', '--ldaptls', action='store_true', help='enable TLS')

    parser.add_argument(
        '--ldap-username-attribute',
        metavar='ATTRIBUTE',
        help=_(
            "The attribute to search username. If no attributes are provided, the default is to use `uid`. It's a good idea to choose an attribute that will be unique across all entries in the subtree you will be using."
        ),
        default='uid',
    )

    parser.add_argument(
        '--ldap-user-filter',
        help=_(
            "search filter to limit LDAP lookup. If not provided, defaults to `(objectClass=*)`, which searches for all objects in the tree. For improved performance it's recommanded to narrow the search to your user object class. e.g.: `(objectClass=posixAccount)`"
        ),
        default='(objectClass=*)',
    )

    parser.add_argument(
        '--ldap-admin-group',
        metavar='CN',
        help=_("list of CN of the group(s) containing administrator. Not cn=groupname or the full DN."),
        action='append',
        default=[],
    )

    parser.add_argument(
        '--ldap-subnet-mgmt-group',
        metavar='CN',
        help=_("list of CN of the group(s) containing Subnet Managers. Not cn=groupname or the full DN."),
        action='append',
        default=[],
    )

    parser.add_argument(
        '--ldap-dnszone-mgmt-group',
        metavar='CN',
        help=_("list of CN of the group(s) containing DNS Zone Managers. Not cn=groupname or the full DN."),
        action='append',
        default=[],
    )

    parser.add_argument(
        '--ldap-user-group',
        metavar='CN',
        help=_("list of CN of the group(s) containing Users. Not cn=groupname or the full DN."),
        action='append',
        default=[],
    )

    parser.add_argument(
        '--ldap-guest-group',
        '--ldap-required-group',
        metavar='CN',
        help=_("list of CN of the group(s) containing Guests. Not cn=groupname or the full DN."),
        action='append',
        default=[],
    )

    parser.add_argument(
        '--ldap-group-filter',
        help=_(
            "search filter to limit LDAP lookup of groups. If not provided, defaults to `(objectClass=*)`, which searches for all objects in the tree. For improved performance it's recommanded to narrow the search to your group object class. e.g.: `(objectClass=posixGroup)`"
        ),
        default='(objectClass=*)',
    )

    parser.add_argument(
        '--ldap-group-attribute',
        metavar='ATTRIBUTE',
        help=_(
            "name of the attribute on the Group that hold the list of members. Default: `member`. Other common value is: `memberUid`"
        ),
        default='member',
    )

    parser.add_argument(
        '--ldap-group-attribute-is-dn',
        help=_("True If the group contains list of user defined with DN instead of username."),
        action='store_true',
    )

    parser.add_argument(
        '--ldap-bind-dn',
        metavar='DN',
        help=_(
            "optional DN used to bind to the server when searching for entries. If not provided, will use an anonymous bind."
        ),
        default="",
    )

    parser.add_argument(
        '--ldap-bind-password',
        metavar='PASSWORD',
        help=_(
            "password to use in conjunction with `--ldap-bind-dn`. Note that the bind password is probably sensitive data, and should be properly protected. You should only use the `--ldap-bind-dn` and `--ldap-bind-password` if you absolutely need them to search the directory."
        ),
        default="",
    )

    parser.add_argument(
        '--ldap-version',
        help=_("version of LDAP in use either 2 or 3. Default to 3."),
        default=3,
        type=int,
        choices=[2, 3],
    )

    parser.add_argument(
        '--ldap-network-timeout',
        metavar='SECONDS',
        help=_("timeout in seconds value used for LDAP connection"),
        default=100,
        type=int,
    )

    parser.add_argument(
        '--ldap-timeout',
        metavar='SECONDS',
        help=_("timeout in seconds value used for LDAP request"),
        default=300,
        type=int,
    )

    parser.add_argument(
        '--ldap-encoding',
        metavar='ENCODING',
        help=_("encoding used by your LDAP server."),
        default="utf-8",
    )

    parser.add_argument(
        '--ldap-fullname-attribute',
        help=_(
            "LDAP attribute for user display name. If `fullname` is blank, the fullname is taken from the `firstname` and `lastname`. Attributes 'cn', or 'displayName' commonly carry full names."
        ),
        default=[],
        action='append',
    )

    parser.add_argument(
        '--ldap-firstname-attribute',
        help=_(
            "LDAP attribute for user first name. Used when the attribute configured for name does not exist. e.g.: `givenName`"
        ),
        default=[],
        action='append',
    )

    parser.add_argument(
        '--ldap-lastname-attribute',
        help=_(
            "LDAP attribute for user last name. Used when the attribute configured for name does not exist. e.g.: `sn`"
        ),
        default=[],
        action='append',
    )

    parser.add_argument(
        '--ldap-email-attribute',
        help=_("LDAP attribute for user email. e.g.: mail, email, userPrincipalName"),
        default=[],
        action='append',
    )

    # Email
    parser.add_argument('--smtp-server', metavar='SERVER', help=_('SMTP server used to send email.'))
    parser.add_argument(
        '--smtp-username', metavar='USERNAME', help=_('Username used to authenticated with the SMTP server.')
    )
    parser.add_argument(
        '--smtp-password', metavar='PASSWORD', help=_('Password used to authenticated with the SMTP server.')
    )
    parser.add_argument('--smtp-from', metavar='SERVER', help=_('Email address used to send email.'))
    parser.add_argument(
        '--smtp-encryption',
        default='none',
        choices=['none', 'ssl', 'starttls'],
        help=_('type of encryption to be used when establishing communication with SMTP server: none, ssl, starttls'),
    )

    parser.add_argument(
        '--notification-catch-all-email',
        metavar='EMAIL',
        help=_('When defined, all notification email will be sent to this email address.'),
        default=None,
    )

    parser.add_argument(
        '--favicon',
        dest='favicon',
        help='location of an icon to be used as a favicon displayed in web browser.',
    )

    parser.add_argument(
        '--header-logo',
        dest='header_logo',
        help='location of an image (preferably a .png) to be used as a replacement for the UDB header logo displayed in navigation bar.',
    )

    # Branding
    parser.add_argument(
        '--header-name',
        metavar='NAME',
        default='Universal Database',
        help=_('Name used in the title for this application.'),
    )
    parser.add_argument(
        '--footer-name',
        metavar='NAME',
        default='UDB',
        help=_('Text displayed in the footer along "power by".'),
    )
    parser.add_argument(
        '--footer-url',
        metavar='URL',
        default="https://gitlab.com/ikus-soft/udb",
        help=_('URL used in the footer along "power by".'),
    )

    parser.add(
        '--password-score',
        type=lambda x: max(1, min(int(x), 4)),
        help=_(
            "Minimum zxcvbn's score for password. Value from 1 to 4. Default value 2. Read more about it here: https://github.com/dropbox/zxcvbn"
        ),
        default=2,
    )

    parser.add(
        '--umask',
        help=_(
            "Force a specific umask value. Usually expressed in octal format, for example, ``0022``. Default value is inherit from parent process."
        ),
        metavar='UMASK',
        type=_umask,
    )

    parser.add(
        '--user',
        help=_(
            "User under which the server will answer requests. In order to use this directive, the process must be run initially as root. If you start the server as a non-root user, it will fail to change to the lesser privileged user, and will instead continue to run as that original user."
        ),
        metavar='USER',
        type=_userid,
    )

    parser.add(
        '--group',
        help=_(
            "Group under which the server will answer requests. In order to use this directive, the process must be run initially as root."
        ),
        metavar='GROUP',
        type=_groupid,
    )

    return parser.parse_args(args, config_file_contents=config_file_contents)
